src/System.Runtime.Extensions/src/System/IO/Path.cs

```diff
@@ -356,58 +357,58 @@ private static string CombineNoChecks(string path1, string path2, string path3)
         {
-            // This routine is optimised to be used with buffs of length 20
-            Debug.Assert(((buff.Length % 5) == 0), "Unexpected hash length");
+            Debug.Assert(bytes != null);
+            Debug.Assert(chars != null);
 
-            // For every 5 bytes, 8 characters are appended.
-            StringBuilder sb = StringBuilderCache.Acquire();
+            // This routine is optimized to be used with bytes of length 9 and chars of length 12.
+            Debug.Assert(byteCount == 9, $"Unexpected {nameof(byteCount)}");
+            Debug.Assert(charCount == 12, $"Unexpected {nameof(charCount)}");
 
-            // Create l char for each of the last 5 bits of each byte.  
+            // For every 5 bytes, 8 characters are appended.
+            // Create l char for each of the last 5 bits of each byte.
             // Consume 3 MSB bits 5 bytes at a time.
-            int len = buff.Length;
-            int i = 0;
-            do
-            {
-                byte b0 = (i < len) ? buff[i++] : (byte)0;
-                byte b1 = (i < len) ? buff[i++] : (byte)0;
-                byte b2 = (i < len) ? buff[i++] : (byte)0;
-                byte b3 = (i < len) ? buff[i++] : (byte)0;
-                byte b4 = (i < len) ? buff[i++] : (byte)0;
-
-                // Consume the 5 Least significant bits of each byte
-                sb.Append(s_Base32Char[b0 & 0x1F]);
-                sb.Append(s_Base32Char[b1 & 0x1F]);
-                sb.Append(s_Base32Char[b2 & 0x1F]);
-                sb.Append(s_Base32Char[b3 & 0x1F]);
-                sb.Append(s_Base32Char[b4 & 0x1F]);
-
-                // Consume 3 MSB of b0, b1, MSB bits 6, 7 of b3, b4
-                sb.Append(s_Base32Char[(
-                        ((b0 & 0xE0) >> 5) |
-                        ((b3 & 0x60) >> 2))]);
-
-                sb.Append(s_Base32Char[(
-                        ((b1 & 0xE0) >> 5) |
-                        ((b4 & 0x60) >> 2))]);
-
-                // Consume 3 MSB bits of b2, 1 MSB bit of b3, b4
-
-                b2 >>= 5;
-
-                Debug.Assert(((b2 & 0xF8) == 0), "Unexpected set bits");
-
-                if ((b3 & 0x80) != 0)
-                    b2 |= 0x08;
-                if ((b4 & 0x80) != 0)
-                    b2 |= 0x10;
-
-                sb.Append(s_Base32Char[b2]);
-
-            } while (i < len);
 
-            return StringBuilderCache.GetStringAndRelease(sb);
+            byte b0 = bytes[0];
+            byte b1 = bytes[1];
+            byte b2 = bytes[2];
+            byte b3 = bytes[3];
+            byte b4 = bytes[4];
+
+            // Consume the 5 Least significant bits of the first 5 bytes
+            chars[0] = s_Base32Char[b0 & 0x1F];
+            chars[1] = s_Base32Char[b1 & 0x1F];
+            chars[2] = s_Base32Char[b2 & 0x1F];
+            chars[3] = s_Base32Char[b3 & 0x1F];
+            chars[4] = s_Base32Char[b4 & 0x1F];
```

Comment:
Nit: As long as you're already modifying all of the consumption sites, consider renaming ```s_Base32Char``` to ```s_base32Char```.